src/xas_tdp_utils.F

```diff
@@ -0,0 +1,3926 @@
+!>        large matrices
+!> amatrix ...
+!> \param amatrix ...
+!> \param bmatrix ...
+!> \param eigenvectors ...
+!> \param eigenvalues ...
+!> \param ilow index of the first eigenvalue to compute
+!> \param iup index of the last eigenvalue to compute
+!> \param work ...
+!> \note Blatantly copied from cp_fm_geeig. The eigenvalues are the first elements of the
+!>       `eigenvalues` array and the eigenvectors are in the first columns of `eigenvectors`
+! **************************************************************************************************
+   SUBROUTINE cp_fm_geeig_syevr(amatrix, bmatrix, eigenvectors, eigenvalues, ilow, iup, work)
+
+      TYPE(cp_fm_type), POINTER                          :: amatrix, bmatrix, eigenvectors
+      REAL(dp), DIMENSION(:)                             :: eigenvalues
+      INTEGER, INTENT(IN)                                :: ilow, iup
+      TYPE(cp_fm_type), POINTER                          :: work
+
+      CHARACTER(len=*), PARAMETER :: routineN = 'cp_fm_geeig_syevr', &
+         routineP = moduleN//':'//routineN
+
+      INTEGER                                            :: nao, neig
+
+      CALL cp_fm_get_info(amatrix, nrow_global=nao)
+      neig = iup-ilow+1
+      ! Cholesky decompose S=U(T)U
+      CALL cp_fm_cholesky_decompose(bmatrix)
+      ! Invert to get U^(-1)
+      CALL cp_fm_triangular_invert(bmatrix)
+      ! Reduce to get U^(-T) * H * U^(-1)
+      CALL cp_fm_triangular_multiply(bmatrix, amatrix, side="R")
+      CALL cp_fm_triangular_multiply(bmatrix, amatrix, transpose_tr=.TRUE.)
+      ! Diagonalize
+      CALL cp_fm_syevr(amatrix, work, eigenvalues, ilow, iup)
+      ! Restore vectors C = U^(-1) * C*
+      CALL cp_fm_triangular_multiply(bmatrix, work)
+      CALL cp_fm_to_fm(work, eigenvectors, neig)
+
+   END SUBROUTINE cp_fm_geeig_syevr
+
+! **************************************************************************************************
+!> \brief  computes selected eigenvalues and, optionally, eigenvectors of
+!>        a real symmetric matrix A distributed in 2D blockcyclic format by
+!>       calling the recommended sequence of ScaLAPACK routines.
+!>
+!> \param matrix ...
+!> \param eigenvectors ...
+!> \param eigenvalues ...
+!> \param ilow ...
+!> \param iup ...
+!> \note  matrix is supposed to be in upper triangular form, and overwritten by this routine
+!>        subsets of eigenvalues/vectors can be selected by
+!>        specifying a range of values or a range of indices for the desired eigenvalues.
+!>        It was once in cp_fm_diag but got removed, moved it back here since it is needed for XAS_TDP
+!>        The eigenvalues/vectors are respectivly the first elements of the array/ columns of the matrix
+!>        Assumes scalapack version 2
+! **************************************************************************************************
+   SUBROUTINE cp_fm_syevr(matrix, eigenvectors, eigenvalues, ilow, iup)
```

Comment:
this does belong (probably together with the above) in `src/fm/cp_fm_diag.F`, but have you seen 2ac9a3ec8c1cf4c53858129cf9a0fa87453e89a6 ?

Reply:
Indeed it does. I moved it here (and cleaned it up a bit) after it was removed from master. I am aware that it crashes from time to time, especially for small matrices. This is also why this is not the default option in my code. 

However, I noticed some significant saving in larger systems where the diagonalization starts being relatively expensive and only a few eigenvalues/eigenvectors are needed.

I would personally would like to keep it. I am happy to move it back to `cp_fm_diag.F` too. We can probably keep it around for those specific application without offering it as an option for overall diagonalization in the code ?